File: competitive-programming/leet-code/423/originalDigits.js
const DIGITS = [
    ["0", 25, [14]],
    ["2", 22, [14]],
    ["4", 20, [5, 14]],
    ["6", 23, [18, 8]],
    ["8", 6, [8, 7]],
    ["5", 5, [8]],
    ["7", 18, []],
    ["3", 7, []],
    ["9", 8, []],
    ["1", 14, []]
]

var originalDigits = function (S) {
    let fmap = new Uint16Array(26),
        ans = new Array(10), len = S.length
    for (let i = 0; i < len; i++)
        fmap[S.charCodeAt(i) - 97]++
    for (let i = 0; i < 10; i++) {
        let [dig, char, rems] = DIGITS[i],
            count = fmap[char]
        for (let j = 0; j < rems.length; j++)
            fmap[rems[j]] -= count
        ans[dig] = dig.repeat(count)
    }
    return ans.join("")
};

File: competitive-programming/leet-code/423/original_digits.py
DIGITS = [
    ["0", 25, [14]],
    ["2", 22, [14]],
    ["4", 20, [5, 14]],
    ["6", 23, [18, 8]],
    ["8", 6, [8, 7]],
    ["5", 5, [8]],
    ["7", 18, []],
    ["3", 7, []],
    ["9", 8, []],
    ["1", 14, []]
]

def original_digits(S: str) -> str:
    fmap = [0] * 26
    ans = [""] * 10
    
    for ch in S:
        fmap[ord(ch) - ord('a')] += 1
    
    for i in range(10):
        dig, char, rems = DIGITS[i]
        count = fmap[char]
        
        for rem in rems:
            fmap[rem] -= count
        
        ans[i] = dig * count
    
    return "".join(ans)